enhancements/aws-pod-identity.md

```diff
@@ -0,0 +1,442 @@
+3. OpenShift should create the OIDC provider in IAM
+4. The token signing CA (`bound-sa-token-signing-certs`) can rotate and should
+   sync between OCP and the OIDC provider in IAM.  Rotation should be seemsless.
+   Assumes applications in pods monitor the token for rotation (AWS SDK does
+   this)
+
+### Non-Goals
+
+1. Convert existing IAM user-based components (e.g. ingress, image-registry,
+   machine-api) to use bound tokens
+
+## Proposal
+
+A new `aws-pod-identity-webhook` image will be added to the release payload
+based on [this](https://github.com/openshift/aws-pod-identity-webhook) repo.
+
+Considerable functionality will be added to the `cloud-credential-operator`
+(CCO).  This functionality is AWS-only and will only be active for clusters
+deployed with `platform: aws`.
+
+- Management of the `aws-pod-identity-webhook` and its dependent resources
+  - Synchronization of the `spec.clientConfig.caBundle` on
+    `MutationWebhookConfiguration` whenever the service serving CA
+    (`service-ca`) rotates
+- Creation of the OIDC provider in IAM (pending on Open Question)
+- Synchronization of the `keys.json` in the S3 bucket pointed to by the OIDC
+  provider with the rotatable public keys contained in
+  `bound-sa-token-signing-certs` configmap in the `openshift-kube-apiserver`
+  namespace (pending on Open Question)
+
+IAM permissions will need to be expanded to include IAM OIDC provider create and
+S3 bucket create/write if the CCO is to manage the OIDC provider (see Open
+Questions section)
+
+In terms of API changes, an additional `serviceAccountRef` field will be added
+to the `CredentialsRequests` CR `spec` alongside the existing [`secretRef`
+field](https://github.com/openshift/cloud-credential-operator/blob/9cc1c1abf898cfa08429d465da0f02d064bd89ee/manifests/00_v1_crd.yaml#L37-L42).
+Usage of this field is an indication by the user that they wish to use this new
+bound token injection mechanism.  The CCO will create a role, instead of a user,
+in response and create a `ServiceAccount` with the appropriate role annotation.
+A pod started with this `ServiceAccount` will have a token injected by the
+mutating webhook.  A `serviceAccountRef` specified on a `CredentialsRequests`
+when the platform is not AWS will result in an error reported in the
+`CredentialsRequests` `status`.
+
+There is also a small installer impact in that we need to change the
+`serviceAccountIssuer` in the `Authentication` config CR day-1 e.g. at install
+manifest render time
+
+### Implementation Details/Notes/Constraints [optional]
+
+This will be deployed automatically and always for OCP on AWS and is inert if
+not explicity utilized by the admin via the `serviceAccountRef` field in the
+`CredentialsRequests` CR.
+
+We might want to abort the deployment of the webhook and disable this whole
+thing if we find that CCO does not have sufficient IAM privileges to take the
+required configuration actions.
+
+#### OIDC Provider Creation and Synchronization
```

Comment:
Can the `ServiceAccountIssuerDiscovery` feature (alpha in 1.18) obviate the need for an s3 bucket?

Reference: https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.18.md#other-api-changes

Reply:
interesting! first i'm hearing of it. let me get up to speed and think on it :+1:

Reply:
Yeah, that would be super nice if the KAS would serve up the discovery and keys json.  Even more amazing if it keeps keys.json in sync with the bound sa signer rotates! (would it @marun?) 

One catch I see is if the KAS is not publicly accessible or only accessible via a bastion or authenticating proxy.

Reply:
> Yeah, that would be super nice if the KAS would serve up the discovery and keys json. Even more amazing if it keeps keys.json in sync when the bound sa signer rotates! (would it @marun?)

Yes, keys.json will be kept in sync. An apiserver's oidc endpoint will serve all the service account public keys it is configured with.

Reply:
I would recommend we serve it as well if not primarily with the option to mirror it to an S3 bucket location via the CCO... could enable other, generic OIDC integrations later and still let us satisfy the disconnected case.

Reply:
As @marun mentioned, this ability is alpha in 1.18 (base for 4.5).

If we want to do this, we have to address the disconnected situation. Short term solution is probably docs on how to continuously mirror the OIDC documents hosted by the KAS.

We would also need to add the feature gate [here](https://github.com/openshift/api/blob/master/config/v1/types_feature.go#L113-L124).

We also take on the risk of it not being fully fleshed out and having to fix things upstream.  The feature seems simple enough so I imagine that risk is low, even though it is alpha.

Reply:
Note that it is possible to specify an external jks uri by supplying `--service-account-jwks-uri` to the apiserver [1]. For a private cluster, this would allow a publicly accessible version of the oidc endpoint to be a mirrored copy while still ensuring that the jks uri resolved to the external rather than internal endpoint. 

1: https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-issuer-discovery (see end of section)